My HashMap Realisation

PavelGozhii opened the pull request.

PavelGozhii committed:
My HashMap Realisation

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    @Override
+    public V remove(K key) {
```
This method is huge, split it to smaller one

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+        V value;
+        Node next;
+
+        public Node getNode(int a) {
```
Constructor should not return null, what is the purpose of it?

PavelGozhii replied:
> Constructor should not return null, what is the purpose of it?

oops, glee code, which I forgot to remove, my guilt

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+
+    @Override
+    public V get(K key) {
+        if (key == null) {
```
I would not recommend you to allow null key, it makes your code more complicated, and actually does not makes much sense

PavelGozhii replied:
> I would not recommend you to allow null key, it makes your code more complicated, and actually does not makes much sense

I read articles on habr, there was such opportunity therefore I realized it, actually it took a lot of time :) though if honestly, I too don't see in it sense

Siril-O replied:
No problem, your code is more complex and you need just to do more job to make it decent and working

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+                    }
+                    while (currNode.key != null) {
+                        currNode = currNode.next;
+                        if (nodesArr[0].key == null) {
```
You have already checked this `if (nodesArr[0].key == null) {` in line 151

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+            }
+
+        } else {
+            int sizeArr = 16;
```
Why 16?

PavelGozhii replied:
> Why 16?

the default value in the Java class, rumor has it that this is a good ratio along from the load factor of 0.75

Siril-O replied:
this is default value for initial size of array. Why do you need sizeArr here, just do simple iteration what is the point to iterate by chucks of 16 ?

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,228 @@
+                }
+            }
+        } else {
+            int sizeArr = 16;
```
Why 16

Siril-O replied:
Same as in remove why iterate in chunks by 16 ?

PavelGozhii committed:
Update MyHashMap.java

PavelGozhii committed:
Update MyHashMap.java

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,196 @@
+        return val;
+    }
+
+    public V getVal(int index, K key) {
```
interface do not declares this method, why it is public ?

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,196 @@
+        System.arraycopy(nodesArr, 0, tempNode, 0, size);
+        this.nodesArr = tempNode;
+        this.size = size;
+        this.threshold = (int) (nodesArr.length * loadFactor);
```
this not needed

PavelGozhii replied:
> this not needed

I do this because I think it makes the code clearer

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,196 @@
+        return val;
+    }
+
+    public V removeVal(int index, K key) {
```
make private

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,196 @@
+        V val = null;
+        if (key != null) {
+            int sizeArr = size;
+            while (sizeArr <= nodesArr.length) {
```
why do you need this while?  just calculate index by hashCode and check if  your bucket contains key

PavelGozhii replied:
> why do you need this while? just calculate index by hashCode and check if your bucket contains key

hash index depends on the size, that is, if for the current size, hashindex is already different for the same key, you need to check all the sizes, maybe there is a better way, but I did not find information about it on the Internet and did not come up with myself

Siril-O replied:
index of bucket is Math.abs(hashCode(key)) % nodesArr.length. just get node by this index and find node with right key if there multiple nodes in bucket.  All your operations should access/remove/modify nodes in this way. When you will add node and you need to make table bigger, just put all data to new array, but be aware that your nodesArr.length will be another so all nodes may have new indexes

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,196 @@
+        V val = null;
+        if (key != null) {
+            int sizeArr = size;
+            while (sizeArr <= nodesArr.length) {
```
same as with get why do you need while ?

PavelGozhii committed:
Update MyHashMap.java

PavelGozhii committed:
Ok, 

but in my humble opinion, it is too primitive and I would not mind to implement a full HashMap in Java

Siril-O commented on MyHashMap.java:
```diff
@@ -0,0 +1,188 @@
+                    currentNode = currentNode.next;
+                    stringValue.append(currentNode.key).append(" ").append(currentNode.value).append("\n");
+                }
+            } catch (Exception e) {
```
do not catch Exception, client of class should do this
